From a2f946ed16ab0e17bf823ec0ec9090efcb44f269 Mon Sep 17 00:00:00 2001
From: shxrrydw <sherrydrv@gmail.com>
Date: Tue, 4 Aug 2026 13:29:18 +0200
Subject: [PATCH] HACK: wintrust: Stub some APIs for some games.

Wine port of the workaround used to play the game in the last years.

App: HK: Star Rail.
---
 dlls/wintrust/crypt.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/dlls/wintrust/crypt.c b/dlls/wintrust/crypt.c
index ad6f264cffb..5e72f879288 100644
--- a/dlls/wintrust/crypt.c
+++ b/dlls/wintrust/crypt.c
@@ -69,6 +69,22 @@ struct catinfo
     WCHAR file[MAX_PATH];
 };
 
+static BOOL needs_cryptstub(void)
+{
+    static int cache = -1;
+
+    const WCHAR *name, *p;
+
+    if (cache != -1) return cache;
+
+    name = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+    if (!name) return FALSE;
+    if ((p = wcsrchr(name, '/'))) name = p + 1;
+    if ((p = wcsrchr(name, '\\'))) name = p + 1;
+    cache = !lstrcmpiW(name, L"StarRail.exe");
+    return cache;
+}
+
 static HCATINFO create_catinfo(const WCHAR *filename)
 {
     struct catinfo *ci;
@@ -470,6 +486,8 @@ HCATINFO WINAPI CryptCATAdminEnumCatalogFromHash(HCATADMIN hCatAdmin, BYTE* pbHa
     static const WCHAR slashW[] = {'\\',0};
     static const WCHAR globW[]  = {'\\','*','.','c','a','t',0};
 
+    static LONG once;
+
     struct catadmin *ca = hCatAdmin;
     WIN32_FIND_DATAW data;
     HCATINFO prev = NULL;
@@ -479,6 +497,11 @@ HCATINFO WINAPI CryptCATAdminEnumCatalogFromHash(HCATADMIN hCatAdmin, BYTE* pbHa
 
     TRACE("%p %p %ld %lx %p\n", hCatAdmin, pbHash, cbHash, dwFlags, phPrevCatInfo);
 
+    if (!once && needs_cryptstub() && !InterlockedExchange(&once, 1))
+    {
+        return (HCATINFO)1;
+    }
+
     if (!ca || ca->magic != CATADMIN_MAGIC || !pbHash || cbHash != 20 || dwFlags)
     {
         SetLastError(ERROR_INVALID_PARAMETER);
@@ -599,11 +622,18 @@ BOOL WINAPI CryptCATAdminReleaseCatalogContext(HCATADMIN hCatAdmin,
                                                HCATINFO hCatInfo,
                                                DWORD dwFlags)
 {
+    static LONG once;
+
     struct catinfo *ci = hCatInfo;
     struct catadmin *ca = hCatAdmin;
 
     TRACE("%p %p %lx\n", hCatAdmin, hCatInfo, dwFlags);
 
+    if (!once && needs_cryptstub() && !InterlockedExchange(&once, 1))
+    {
+        return 1;
+    }
+
     if (!ca || ca->magic != CATADMIN_MAGIC || !ci || ci->magic != CATINFO_MAGIC)
     {
         SetLastError(ERROR_INVALID_PARAMETER);
@@ -1022,10 +1052,17 @@ out:
  */
 BOOL WINAPI CryptCATCatalogInfoFromContext(HCATINFO hcatinfo, CATALOG_INFO *info, DWORD flags)
 {
+    static LONG once;
+
     struct catinfo *ci = hcatinfo;
 
     TRACE("%p, %p, %lx\n", hcatinfo, info, flags);
 
+    if (!once && needs_cryptstub() && !InterlockedExchange(&once, 1))
+    {
+        return 1;
+    }
+
     if (!hcatinfo || hcatinfo == INVALID_HANDLE_VALUE || ci->magic != CATINFO_MAGIC ||
         flags || !info || info->cbStruct != sizeof(*info))
     {
